### ISSUE TITLE
Job never run on some devices

### ISSUE THREAD

**Issue by andygeers** [firebase/firebase-jobdispatcher-android#204]
I am using Firebase JobDispatcher to run a scheduled task that requires network access. It's working perfectly for me, and I know it's working well for many users too. But having released my app, a meaningful number of users are reporting that it's not working - and debug info suggests that the task is just never getting run. I just don't know where to even begin in debugging this issue remotely - if I could reproduce it I would look at the schedule to see if the job was listed, but I presume that's not possible from a distance / via code.

I set it the job up like so:

```
FirebaseJobDispatcher dispatcher = new FirebaseJobDispatcher(new GooglePlayDriver(context));
Job.Builder builder = dispatcher.newJobBuilder()
                //persist the task across boots
                .setLifetime(Lifetime.FOREVER)
                //call this service when the criteria are met.
                .setService(MyJobService.class)
                //unique id of the task
                .setTag("com.andygeers.project.JobService")
                //overwrite an existing job with the same tag
                .setReplaceCurrent(true)
                // retry with exponential backoff
                .setRetryStrategy(RetryStrategy.DEFAULT_LINEAR)        
                //Run this job only when the network is available.
                .setConstraints(Constraint.ON_ANY_NETWORK);
                // Other versions have recurring=true and trigger as Trigger.executionWindow(60 * 6, 60 * 60)
                .setRecurring(false)    
                .setTrigger(Trigger.NOW);

Job job = builder.build();
```

Then I have a JobService method like so, but this is apparently never run, according to my debug info: (whereas on my own device, that same debug info shows that it does run just fine)
```

public static boolean mHasStartedSyncJob;

@Override
    public boolean onStartJob(final JobParameters params) {
        // Debug code - on working devices, this reports true. On broken devices,
        // this remains false.
        mHasStartedSyncJob = true;

        //Offloading work to a new thread.
        new Thread(new Runnable() {
            @Override
            public void run() {
                syncFeeds(params);
            }
        }).start();

        return true; // Answers the question: "Is there still work going on?"
}
```

One example device that doesn't work:
Android SDK Version: 24
Device: a5y17ltecan / SM-A520W / a5y17ltevl / aarch64

Any thoughts on where to even begin with debugging this?

**Comment by jplazcano87** [comment]
maybe is the google play services version in the emulator / device

**Comment by andygeers** [comment]
He checked in Google Play and it didn't say there were any updates available for Google Play Services - how could I tell?

**Comment by jplazcano87** [comment]
i'm having the same problem, and no help, is sad because this library appears in the android official docs, but it's work only in some selected devices and apis, is so unreliable.

**Comment by jplazcano87** [comment]
still no word on this issue?

**Comment by andygeers** [comment]
As one extra data point: one user who had consistently had this problem over a period of several weeks did a factory reset of his phone, and then it started working again without me changing my app. Obviously I can't ask all my users to do this - would love to know what made the difference!

**Comment by nycodes9** [comment]
try replacing `Trigger.NOW` with `Trigger.executionWindow(0, 0)`

**Comment by andygeers** [comment]
Would that change explain why it works for most users? (it's a bit hard for me to "try" much now that the main beta user I had who was tech-savvy enough to try things for me has done a factory reset on his phone and the issue has gone away - but I have plenty of other less technical users complaining that the issue is ongoing)

**Comment by jplazcano87** [comment]
if you really need a job to run, in a critical part of your app, don't use this library, is broken, jobs only run in some devices, no one helps you.

**Comment by salmancs14** [comment]
firstly it has power mode issue, even if you handle that firebase dispatcher not working in some devices.

**Comment by hivian** [comment]
Just update your firebase to the latest version (0.8.5) It solved the problem for me.

implementation 'com.firebase:firebase-jobdispatcher:0.8.5'

**Comment by salmancs14** [comment]
@hivian after updating to 0.8.5, still not scheduling in some devices and it is not even throwing `ScheduleFailedException`

**Comment by jplazcano87** [comment]
@salmancs14 like i said early, do not use this lib for any critical work. use alarm manager, sync adapter or other native and backwards compatible api, if you need a task to run in every single device, also keep in mind the background execution limits in android Oreo

**Comment by salmancs14** [comment]
@jplazcano87 you need to schedule a job in android Oreo that is why we need to shift to job dispatcher. Don't use Alarm Manager

**Comment by jplazcano87** [comment]
@salmancs14 jobs are native in newer versions of android, this library is not needed for those versions, i'm fully aware of the background execution limits in oreo.

**Comment by Syntey** [comment]
I'm also having this exact problem. It's working for me, but for 2 of my testers that are on Android 6 it's not working.

**Comment by tomasmichael995** [comment]
It's the first time that I've tried to use this library. Does not work on a Samsung and a Huawai device. The job never starts and I did turn of the optimalization for my app.

**Comment by a23sokolov** [comment]
Jobs scheduled but not started on Emulator API 19, 21. And appropriate work on Api 26? As a result i think @jplazcano87 right don't use this lib in project

**Comment by CatalinPrata** [comment]
I have the same experience on a Samsung S6 with Android OS 7, the service gets triggered only after about 30 minutes or so but sometimes it gets triggered instantly.

dispatcher.newJobBuilder()
                        .setService(MyService.class)
                        .setLifetime(Lifetime.FOREVER)
                        .setTag(A_TAG)
                        .setExtras(myBundle)
                        .build();

I tried with Trigger.NOW and Trigger.executionWindow(0, 0) but no luck.

**Comment by lordplagus02** [comment]
You'll notice in general there are issues with the trigger execution time, which is making the library quite unreliable (Google, your docs recommend this, where are you?). I believe once the job triggering is fixed, all the reliability issues will disappear.

**Comment by hivian** [comment]
Meanwhile, you can try Evernote Android Job.

**Comment by pzmudzinski** [comment]
I've seen this issue open and decided not to use this library and went with plain `JobIntentService`.

**Comment by theernis456** [comment]
I have figured out that it does not work on some devices with power saving mode on,  turning off power saving mode schedules job normally.

**Comment by satheeshwaran** [comment]
I am facing issues on a One plus 3 running Oreo. Should I use this or just implement JobScheduler?

**Comment by jplazcano87** [comment]
Oreo has the JobScheduler api included, so there's no need to use another external library

**Comment by VillvaTharisan19** [comment]
I've and issue with setService with MyJobService.class. Need solution for this?

Error: setService (java.lang.Class <? extends com.firebase.jobdispatcher.Jobservice>)

private void scheduleJob() { // Under FirebaseServiceMessaging class
       
        FirebaseJobDispatcher dispatcher = new FirebaseJobDispatcher(new GooglePlayDriver(this));
        Job myJob = dispatcher.newJobBuilder()
                **.setService(MyJobService.class) // This Line (error)**
                .setTag("my-job-tag")
                .build();
        dispatcher.schedule(myJob);
       
    }

public class MyJobService extends JobService { // MyJobService class

    @Override
    public boolean onStartJob(JobParameters job) {
        // Do some work here

        return false; // Answers the question: "Is there still work going on?"
    }

    @Override
    public boolean onStopJob(JobParameters job) {
        return false; // Answers the question: "Should this job be retried?"
    }

}

**Comment by recheej** [comment]
@VillvaTharisan19 what exactly is the error? Post the stack trace.

**Comment by RyanDawkins** [comment]
The scenario I'm currently working through:

Version: 0.8.5
Trigger: executionWindow(0,0)
Constraints: ANY_NETWORK

The way I'm testing is having my app schedule the job while mobile data is off. Then after job is scheduled, turn on mobile data.

**Comment by kpradeepkumarreddy** [comment]
> I am facing issues on a One plus 3 running Oreo. Should I use this or just implement JobScheduler?

i'm also facing the same issue. did you find any solution ?

**Comment by kpradeepkumarreddy** [comment]
Most of the times the issue could be because of wrong parameters used for this API.  In my case i was setting setReplaceCurrent() parameter as true. so i think it was getting scheduled for next day or so. Let me see, now i modified that parameter to false. And i also have unique tag for the job.

**Comment by Shahar2k5** [comment]
Any update on that?
Is this library production safe?
I'm updating my app, replacing services with jobs.
would love to get some insights.

**Comment by Shahar2k5** [comment]
@andygeers @jplazcano87 @nycodes9 @salmancs14 @hivian 
Any updates regarding that issue?
Is this library production safe?
I'm updating my app, replacing services with jobs.
would love to get some insights.

**Comment by Shahar2k5** [comment]
> Most of the times the issue could be because of wrong parameters used for this API. In my case i was setting setReplaceCurrent() parameter as true. so i think it was getting scheduled for next day or so. Let me see, now i modified that parameter to false. And i also have unique tag for the job.
Any updates?

**Comment by jplazcano87** [comment]
> @andygeers @jplazcano87 @nycodes9 @salmancs14 @hivian
> Any updates regarding that issue?
> Is this library production safe?
> I'm updating my app, replacing services with jobs.
> would love to get some insights.

Work Manager is the way to go, if you are replacing services with jobs, i recommend to watch this https://www.youtube.com/watch?v=IrKoBFLwTN0 is a talk from the 2018 google i/o

**Comment by Shahar2k5** [comment]
> > @andygeers @jplazcano87 @nycodes9 @salmancs14 @hivian
> > Any updates regarding that issue?
> > Is this library production safe?
> > I'm updating my app, replacing services with jobs.
> > would love to get some insights.
> 
> Work Manager is the way to go https://developer.android.com/topic/libraries/architecture/workmanager/ , if you are replacing services with jobs, i recommend to watch this https://www.youtube.com/watch?v=IrKoBFLwTN0 is a talk from the 2018 google i/o

True, but WorkManager is in alpha, I think that's a bit hardcore for production.
Every release they have an API breaking changes

**Comment by Leslio7** [comment]
In my case, the device's "Power saving" feature was the culprit. Check if it's enabled and disable it.

**Comment by samtstern** [comment]
In April 2019 we announced that Firebase Job Dispatcher would be deprecated today, April 7th 2020.  For this reason we are going to close all open issues and archive the repository.  We will also stop supporting FJD for new app installations (such as those targeting sdk versions greater than Android Q).  Existing apps that rely on FJD should continue to function as usual.

While we will no longer work on FJD, we encourage you to [migrate to Work Manager](https://developer.android.com/topic/libraries/architecture/workmanager/migrating-fb) which is a superior tool and fully supported by the Android team.

Thank you for taking the time to try the library and provide this feedback.  We sincerely appreciate your effort and we apologize for not addressing it in time.
